lib/ansible/modules/network/ios/ios_ping.py

```diff
@@ -0,0 +1,227 @@
+from ansible.module_utils.connection import exec_command
+from ansible.module_utils.ios import run_commands
+from ansible.module_utils.ios import ios_argument_spec, check_args
+import re
+
+
+def main():
+    """ main entry point for module execution
+    """
+    argument_spec = dict(
+        count=dict(type="int"),
+        dest=dict(type="str", required=True),
+        source=dict(type="str"),
+        state=dict(type="str", choices=["absent", "present"], default="present"),
+        vrf=dict(type="str")
+    )
+
+    argument_spec.update(ios_argument_spec)
+
+    module = AnsibleModule(argument_spec=argument_spec)
+
+    count = module.params["count"]
+    dest = module.params["dest"]
+    source = module.params["source"]
+    vrf = module.params["vrf"]
+
+    warnings = list()
+    check_args(module, warnings)
+
+    results = {}
+    if warnings:
+        results["warnings"] = warnings
+
+    results["updates"] = [build_ping(dest, count, source, vrf)]
+
+    ping_results = run_commands(module, commands=results["updates"])
+    ping_results_list = ping_results[0].split("\n")
+    results["action"] = ping_results_list[1]
+
+    success, rx, tx, rtt = parse_ping(ping_results_list[3])
+    loss = abs(100 - int(success))
+    results["packet_loss"] = str(loss) + "%"
+    results["packets_rx"] = rx
+    results["packets_tx"] = tx
+    results["count"] = tx
+    results["rtt"] = rtt
+
+    validate_results(module, loss, results)
+
+
+def build_ping(dest, count=None, source=None, vrf=None):
+    """
+    Function to build the command to send to the terminal for the switch
+    to execute. All args come from the module's unique params.
+    """
+    if vrf is not None:
+        cmd = "ping {} {}".format(vrf, dest)
+    else:
+        cmd = "ping {}".format(dest)
+
+    if count is not None:
+        cmd += " repeat {}".format(str(count))
+
+    if source is not None:
+        cmd += " source {}".format(source)
+
+    return cmd
+
+
+def parse_ping(ping_stats):
+    """
+    Function used to parse the statistical information from the ping response.
+    Example: "Success rate is 100 percent (5/5), round-trip min/avg/max = 1/2/8 ms"
+    Returns the percent of packet loss, recieved packets, transmitted packets, and RTT dict.
+    """
+    rate_re = re.compile("^\w+\s+\w+\s+\w+\s+(?P<pct>\d+)\s+\w+\s+\((?P<rx>\d+)\/(?P<tx>\d+)\)")
+    rtt_re = re.compile(".*,\s+\S+\s+\S+\s+=\s+(?P<min>\d+)\/(?P<avg>\d+)\/(?P<max>\d+)\s+\w+\s*$|.*\s*$")
+
+    rate = rate_re.match(ping_stats)
+    rtt = rtt_re.match(ping_stats)
+
+    return rate.group("pct"), rate.group("rx"), rate.group("tx"), rtt.groupdict()
+
+
+def validate_results(module, loss, results):
+    """
+    This function is used to validate whether the ping results were expected per "state" param.
+    """
+    state = module.params["state"]
+    if state == "present" and loss != 100:
+        module.exit_json(msg="Ping succeeded expectedly", **results)
```

Comment:
`msg` is not in your `RETURN` documentation, but that's okay because its message is implied by the module succeeding. This whole section could more clearly be put as follows:

```
if state == 'present' and loss == 100:
    module.fail_json('Ping failed unexpectedly', **results)
elif state == 'absent' and loss != 100:
    module.fail_json('Ping succeeded unexpectedly', **results)

module.exit_json(**results)